v2/docker/schema2/manifest.go

```diff
@@ -1,23 +1,76 @@
 package schema2
 
 import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+
 	"github.com/docker/distribution"
+	"github.com/docker/distribution/manifest/manifestlist"
 	v2 "github.com/docker/distribution/manifest/schema2"
+	"github.com/docker/distribution/reference"
+	"github.com/grafeas/voucher/v2/docker/uri"
 )
 
 // IsManifest returns true if the passed manifest is a schema2 manifest.
 func IsManifest(m distribution.Manifest) bool {
-	_, ok := m.(*v2.DeserializedManifest)
-	return ok
+	switch m.(type) {
+	case *v2.DeserializedManifest, *manifestlist.DeserializedManifestList:
+		return true
+	default:
+		return false
+	}
 }
 
 // ToManifest casts a distribution.Manifest to a schema2.Manifest. It panics
 // if it passed anything other than a schema2.DeserialzedManifest.
-func ToManifest(manifest distribution.Manifest) v2.Manifest {
-	schema2Manifest, ok := manifest.(*v2.DeserializedManifest)
-	if !ok {
-		panic("schema2.ToManifest was passed a non-schema2.DeserializedManifest")
+func ToManifest(client *http.Client, ref reference.Named, manifest distribution.Manifest) (v2.Manifest, error) {
+	switch m := manifest.(type) {
+	case *v2.DeserializedManifest:
+		return m.Manifest, nil
+	case *manifestlist.DeserializedManifestList:
+		return resolveManifestFromList(client, ref, m)
+	default:
+		return v2.Manifest{}, fmt.Errorf("schema2.ToManifest was passed a %T", manifest)
 	}
+}
 
-	return schema2Manifest.Manifest
+// Ugly method to override the target os/arch without wiring the voucher config to this context
+var targetOS, targetArch string
+
+func init() {
+	targetOS = os.Getenv("VOUCHER_TARGET_OS")
+	if targetOS == "" {
+		targetOS = "linux"
+	}
+	targetArch = os.Getenv("VOUCHER_TARGET_ARCH")
+	if targetArch == "" {
+		targetArch = "amd64"
+	}
+}
+
+func resolveManifestFromList(client *http.Client, ref reference.Named, mfs *manifestlist.DeserializedManifestList) (v2.Manifest, error) {
+	for _, mf := range mfs.Manifests {
+		if mf.Platform.Architecture != targetArch || mf.Platform.OS != targetOS {
+			continue
+		}
+
+		req, err := http.NewRequest("GET", uri.GetManifestURI(ref, string(mf.Digest)), nil)
+		if err != nil {
+			return v2.Manifest{}, fmt.Errorf("preparing request to fetch manifest from list: %w", err)
+		}
+		resp, err := client.Do(req)
+		if err != nil {
+			return v2.Manifest{}, fmt.Errorf("fetching manifest from list: %w", err)
+		}
+		defer resp.Body.Close()
+
+		var archManifest v2.DeserializedManifest
+		if err := json.NewDecoder(resp.Body).Decode(&archManifest); err != nil {
```

Comment:
interesting, I didn't know json can do this + properly decode into the type we wanted

Reply:
Yep! the [`json:""` struct fields](https://github.com/distribution/distribution/blob/6affafd1f030087d88f88841bf66a8abe2bf4d24/manifest/schema2/manifest.go#L63-L72) define the mapping and are a good signal the struct wants to be json-ed.

The fields of `DeserializedManifest` don't add anything - a `v2.Manifest` would work just as well here, but the file was already using `DeserializedManifest` so I stuck with that 🤷 . We could potentially save some allocations here.